bundles/alpha.codegen/src/alpha/codegen/writeC/ReduceExprConverter.xtend

```diff
@@ -0,0 +1,224 @@
+package alpha.codegen.writeC
+
+import alpha.codegen.Expression
+import alpha.codegen.Factory
+import alpha.codegen.FunctionBuilder
+import alpha.codegen.MacroStmt
+import alpha.codegen.ProgramBuilder
+import alpha.codegen.isl.ASTConverter
+import alpha.codegen.isl.LoopGenerator
+import alpha.model.ReduceExpression
+import fr.irisa.cairn.jnimap.isl.ISLAff
+import fr.irisa.cairn.jnimap.isl.ISLConstraint
+import fr.irisa.cairn.jnimap.isl.ISLDimType
+import fr.irisa.cairn.jnimap.isl.ISLSpace
+import java.util.HashSet
+
+/**
+ * Converts an Alpha reduce expression into the appropriate C AST nodes.
+ * A new function is created and added to the program which computes the reduction,
+ * and the appropriate function call expression is returned.
+ */
+class ReduceExprConverter {
+	/**
+	 * A counter for the number of reductions that have been created.
+	 * This is used for determining the names of functions and macros
+	 * which will be emitted.
+	 */
+	protected static var nextReductionId = 0
+	
+	/** The name of the reduction variable inside of reduce functions. */
+	protected static val reduceVarName = "reduceVar"
+	
+	/**
+	 * Converts an Alpha reduce expression into the appropriate C AST nodes.
+	 * A new function is created and added to the program which computes the reduction,
+	 * and the appropriate function call expression is returned.
+	 */
+	def static Expression convertExpr(ProgramBuilder program, ReduceExpression expr) {
+		// Create the reduce function and add it to the program.
+		val reduceFunction = createReduceFunction(program, expr)
+		program.addFunction(reduceFunction)
+		
+		// Return a call to the reduce function.
+		val callArguments = #[expr.contextDomain.paramNames, expr.contextDomain.indexNames].flatten
+		return Factory.callExpr(reduceFunction.name, callArguments)
+	}
+	
+	/** Creates the function which evaluates the reduction at a specific output point. */
+	def protected static createReduceFunction(ProgramBuilder program, ReduceExpression expr) {
+		// Capture the next reduction ID so we can create function and macro names.
+		val myId = nextReductionId.toString
+		nextReductionId += 1
+		
+		// Start building the reduce function.
+		val reduceFunctionName = "reduce" + myId
```

Comment:
It’s possible for the input alpha program to have a parameter called “reduce0” which will generate the following:
```
float reduce0(long N, long reduceVar, long ip) {
	float reduceVar;
	long j;
```
which won’t compile.

At this point, it looks like all of the variables (and their names) are available in `program`. Could we add a check here that `reduceFunctionName` does not match other variable names in the program? And if so, just repeatedly prepend underscores while it’s not unique? Or perhaps prefix all generated names with “__”? It’d still be possible for “__reduceVar” to be used in the input program, but it’s probably less likely.

Reply:
I see the `makeParameterName` which does exactly this for the index param names. I tried adding “ip” as a parameter name in addition to “reduceVar” and it correctly generates “ipp”. Maybe we can just do the same thing with a function like `makeReduceFunctionName`.

Reply:
Do we want to just check for uniqueness here, or more globally? For example, we could also have an Alpha variable called `min`, which would conflict with the `min` function macro.

Reply:
Ideally, globally. At any point where names are generated, there should be a check that it does not shadow/conflict with any other names in scope. We should probably be maintaining a true “symbol table” for this.